nodeup/pkg/model/containerd.go

```diff
@@ -0,0 +1,415 @@
+		}
+	}
+
+	// Split into major.minor.(patch+pr+meta)
+	parts := strings.SplitN(containerdVersion, ".", 3)
+	if len(parts) != 3 {
+		return fmt.Errorf("error parsing containerd version %q, no Major.Minor.Patch elements found", containerdVersion)
+	}
+
+	// Validate major
+	containerdVersionMajor, err := strconv.ParseInt(parts[0], 10, 64)
+	if err != nil {
+		return fmt.Errorf("error parsing major containerd version %q: %v", parts[0], err)
+	}
+
+	// Validate minor
+	containerdVersionMinor, err := strconv.ParseInt(parts[1], 10, 64)
+	if err != nil {
+		return fmt.Errorf("error parsing minor containerd version %q: %v", parts[1], err)
+	}
+
+	c.AddTask(b.buildSystemdService(containerdVersionMajor, containerdVersionMinor))
+
+	if err := b.buildSysconfig(c); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *ContainerdBuilder) buildSystemdService(containerdVersionMajor int64, containerdVersionMinor int64) *nodetasks.Service {
+	manifest := &systemd.Manifest{}
+	manifest.Set("Unit", "Description", "containerd container runtime")
+	manifest.Set("Unit", "Documentation", "https://containerd.io")
+	manifest.Set("Unit", "After", "network.target local-fs.target")
+
+	manifest.Set("Service", "EnvironmentFile", "/etc/sysconfig/containerd")
+	manifest.Set("Service", "EnvironmentFile", "/etc/environment")
+	manifest.Set("Service", "ExecStartPre", "-/sbin/modprobe overlay")
+	manifest.Set("Service", "ExecStartPre", "-/sbin/modprobe overlay")
+	manifest.Set("Service", "ExecStart", "/usr/bin/containerd -c /etc/containerd/config-kops.toml \"$CONTAINERD_OPTS\"")
+
+	// kill only the containerd process, not all processes in the cgroup
+	manifest.Set("Service", "KillMode", "process")
+	// set delegate yes so that systemd does not reset the cgroups of containerd containers
+	manifest.Set("Service", "Delegate", "yes")
+
+	manifest.Set("Service", "LimitNOFILE", "1048576")
+	manifest.Set("Service", "LimitNPROC", "infinity")
+	manifest.Set("Service", "LimitCORE", "infinity")
+	manifest.Set("Service", "TasksMax", "infinity")
+
+	manifest.Set("Service", "Restart", "always")
+	manifest.Set("Service", "RestartSec", "2s")
+	manifest.Set("Service", "StartLimitInterval", "0")
+	manifest.Set("Service", "TimeoutStartSec", "0")
+
+	manifest.Set("Install", "WantedBy", "multi-user.target")
+
+	manifestString := manifest.Render()
+	klog.V(8).Infof("Built service manifest %q\n%s", "containerd", manifestString)
+
+	service := &nodetasks.Service{
+		Name:       "containerd.service",
+		Definition: s(manifestString),
+	}
+
+	service.InitDefaults()
+
+	return service
+}
+
+// buildContainerOSConfigurationDropIn is responsible for configuring the containerd daemon options
+func (b *ContainerdBuilder) buildContainerOSConfigurationDropIn(c *fi.ModelBuilderContext) error {
+	lines := []string{
+		"[Service]",
+		"EnvironmentFile=/etc/sysconfig/containerd",
+		"EnvironmentFile=/etc/environment",
+	}
+
+	if b.IsKubernetesGTE("1.10") {
```

Comment:
We have containerd as separate since 1.16. This change removes containerd dependency from Docker packages which is not ideal if we want to keep this 1.18+.

My idea would be to add back containerd dependency to Docker packages and just add more section for Docker 19.03.5 that would be the default in 1.18 and would have containerd dependency handled by the new code.